Atomic/scan.py

```diff
@@ -517,3 +526,33 @@ def write_persistent_data(self):
 
         with open(summary_file, 'w') as f:
             json.dump(persistent_data, f, indent=4)
+
+    def harden(self):
+        for target in self.scan_list:
+            util.write_out("Hardening target {}.\n".format(target.id))
+            temp_dir = tempfile.mkdtemp()
+            fix_script = os.path.join(self.results_dir, target.id, "fix.sh")
+
+            shutil.copy(fix_script, temp_dir)
+            dockerfile_path = os.path.join(temp_dir, "Dockerfile")
+            with open(dockerfile_path, "w") as f:
+                f.write("FROM " + target.id + "\n")
+                f.write("COPY fix.sh /\n")
+                f.write("RUN chmod +x /fix.sh; /fix.sh\n")
+            docker_build_cmd = ['docker', 'build', '--no-cache', temp_dir]
+            try:
+                build_process = subprocess.run(
+                        docker_build_cmd,
+                        check=True,
+                        stdout=subprocess.PIPE,
+                        stderr=subprocess.PIPE,
+                    )
+            except subprocess.CalledProcessError as e:
+                util.write_err("Cannot build hardened image from {}. Docker returned {}.\n".format(target.id, e.returncode))
+                util.write_err("Details:\n{}\n".format(e.stderr.decode("utf-8")))
+                continue
+            build_output = build_process.stdout.decode("utf-8")
+            image_id = build_output.splitlines()[-1].split()[-1]
+            util.write_out("Sucessfully built hardened image {} from {}.\n".format(image_id, target.id))
+            shutil.rmtree(temp_dir)
```

Comment:
This should absolutely be in `finally` and everything that can throw an exception after we create it via mkdtemp has to be in `try`. We need to make sure this is cleaned up.